File: atlas/strategies.py
from typing import List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum
from metadata import metadata, TraceEntry
from data_retriever import data_retriever, ColumnChunk

class PlacementStrategy(Enum):
    HOT_TABLE = 1
    HOT_COLUMN = 2
    LOPT = 3

@dataclass
class Placement:
    column_id: int
    device_id: int

def hot_table_strategy() -> List[Placement]:
    table_access_counts = {}
    for entry in metadata.trace:
        table_access_counts[entry.table_id] = table_access_counts.get(entry.table_id, 0) + 1
    
    sorted_tables = sorted(table_access_counts.items(), key=lambda x: x[1], reverse=True)
    sorted_devices = sorted(metadata.devices.values(), key=lambda d: d.cost_per_gb, reverse=True)
    
    placements = []
    for (table_id, _), device in zip(sorted_tables, sorted_devices):
        table = metadata.get_table(table_id)
        placements.extend([Placement(col.id, device.id) for col in table.columns])
    
    return placements

def hot_column_strategy() -> List[Placement]:
    column_access_counts = {}
    for entry in metadata.trace:
        for col_id in entry.column_ids:
            column_access_counts[col_id] = column_access_counts.get(col_id, 0) + 1
    
    sorted_columns = sorted(column_access_counts.items(), key=lambda x: x[1], reverse=True)
    sorted_devices = sorted(metadata.devices.values(), key=lambda d: d.cost_per_gb, reverse=True)
    
    return [Placement(col_id, device.id) for (col_id, _), device in zip(sorted_columns, sorted_devices)]

def lopt_strategy(budget: Optional[float]) -> List[Placement]:
    column_sizes = {col_id: col.size for col_id, col in metadata.columns.items()}
    sorted_columns = sorted(column_sizes.items(), key=lambda x: x[1], reverse=True)
    sorted_devices = sorted(metadata.devices.values(), key=lambda d: d.cost_per_gb)
    
    placements = []
    total_cost = 0
    for col_id, size in sorted_columns:
        for device in sorted_devices:
            cost = (size / (1024 * 1024 * 1024)) * device.cost_per_gb
            if budget is None or total_cost + cost <= budget:
                placements.append(Placement(col_id, device.id))
                total_cost += cost
                break
    
    return placements

def place_data(strategy: PlacementStrategy, budget: Optional[float] = None) -> List[Placement]:
    if strategy == PlacementStrategy.HOT_TABLE:
        return hot_table_strategy()
    elif strategy == PlacementStrategy.HOT_COLUMN:
        return hot_column_strategy()
    elif strategy == PlacementStrategy.LOPT:
        return lopt_strategy(budget)

def predict_scan_time(placements: List[Placement], table_scan: TraceEntry) -> float:
    max_device_time = 0
    devices_used = {}
    
    for col_id in table_scan.column_ids:
        placement = next(p for p in placements if p.column_id == col_id)
        device = metadata.get_device(placement.device_id)
        device_model = metadata.get_device_model(placement.device_id)
        column = metadata.get_column(col_id)
        
        compression_throughput = {
            CompressionAlgorithm.NONE: device_model.none_throughput,
            CompressionAlgorithm.LZ4: device_model.lz4_throughput,
            CompressionAlgorithm.ZSTD: device_model.zstd_throughput
        }[device.compression]
        
        column_time = device_model.seek_time + (column.size / compression_throughput)
        
        if placement.device_id not in devices_used:
            devices_used[placement.device_id] = column_time
        else:
            devices_used[placement.device_id] += column_time
    
    return max(devices_used.values())

def predict_total_time(placements: List[Placement], trace: List[TraceEntry]) -> float:
    return sum(predict_scan_time(placements, table_scan) for table_scan in trace)

def calculate_total_cost(placements: List[Placement]) -> float:
    return sum(
        metadata.get_device(p.device_id).cost_per_gb * 
        (metadata.get_column(p.column_id).size / (1024 * 1024 * 1024))
        for p in placements
    )

def optimize_placement(budget: float) -> List[Placement]:
    placements = place_data(PlacementStrategy.LOPT, budget)
    total_time = predict_total_time(placements, metadata.trace)
    total_cost = calculate_total_cost(placements)
    print(f"Predicted total time: {total_time:.2f} s")
    print(f"Total cost: {total_cost:.2f}")
    return placements

def apply_placements(placements: List[Placement]):
    for placement in placements:
        column = metadata.get_column(placement.column_id)
        target_device = metadata.get_device(placement.device_id)
        chunks = data_retriever.column_chunks.get(placement.column_id, [])
        
        for chunk in chunks:
            if chunk.device_id != placement.device_id:
                data_retriever.move_column_chunk(chunk, placement.device_id)
        
        print(f"Moved column {column.name} to device {target_device.name}")

File: epaxos/src/strategies.rs
use std::collections::HashMap;
use crate::metadata::{Metadata, TraceEntry};
use crate::data_retriever::DataRetriever;
use crate::kv_store::KVStore;

#[derive(Debug, Clone, PartialEq)]
pub enum PlacementStrategy {
    HOT_TABLE,
    HOT_COLUMN,
    LOPT,
}

#[derive(Debug, Clone)]
pub struct Placement {
    pub column_id: u32,
    pub device_id: u32,
}

pub fn hot_table_strategy(metadata: &Metadata) -> Vec<Placement> {
    let mut table_access_counts: HashMap<u32, u32> = HashMap::new();
    for entry in &metadata.trace {
        *table_access_counts.entry(entry.table_id).or_insert(0) += 1;
    }

    let mut sorted_tables: Vec<_> = table_access_counts.into_iter().collect();
    sorted_tables.sort_by(|a, b| b.1.cmp(&a.1));

    let mut sorted_devices: Vec<_> = metadata.devices.values().collect();
    sorted_devices.sort_by(|a, b| b.cost_per_gb.partial_cmp(&a.cost_per_gb).unwrap());

    let mut placements = Vec::new();
    for ((table_id, _), device) in sorted_tables.iter().zip(sorted_devices.iter()) {
        if let Some(table) = metadata.get_table(*table_id) {
            for column in &table.columns {
                placements.push(Placement {
                    column_id: column.id,
                    device_id: device.id,
                });
            }
        }
    }

    placements
}

pub fn hot_column_strategy(metadata: &Metadata) -> Vec<Placement> {
    let mut column_access_counts: HashMap<u32, u32> = HashMap::new();
    for entry in &metadata.trace {
        for &col_id in &entry.column_ids {
            *column_access_counts.entry(col_id).or_insert(0) += 1;
        }
    }

    let mut sorted_columns: Vec<_> = column_access_counts.into_iter().collect();
    sorted_columns.sort_by(|a, b| b.1.cmp(&a.1));

    let mut sorted_devices: Vec<_> = metadata.devices.values().collect();
    sorted_devices.sort_by(|a, b| b.cost_per_gb.partial_cmp(&a.cost_per_gb).unwrap());

    sorted_columns.iter().zip(sorted_devices.iter().cycle()).map(|((col_id, _), device)| {
        Placement {
            column_id: *col_id,
            device_id: device.id,
        }
    }).collect()
}

pub fn lopt_strategy(metadata: &Metadata, budget: Option<f32>) -> Vec<Placement> {
    let mut column_sizes: Vec<_> = metadata.columns.iter().map(|(_, col)| (col.id, col.size)).collect();
    column_sizes.sort_by(|a, b| b.1.cmp(&a.1));

    let mut sorted_devices: Vec<_> = metadata.devices.values().collect();
    sorted_devices.sort_by(|a, b| a.cost_per_gb.partial_cmp(&b.cost_per_gb).unwrap());

    let mut placements = Vec::new();
    let mut total_cost = 0.0;

    for (col_id, size) in column_sizes {
        for device in &sorted_devices {
            let cost = (size as f32 / (1024.0 * 1024.0 * 1024.0)) * device.cost_per_gb;
            if budget.is_none() || total_cost + cost <= budget.unwrap() {
                placements.push(Placement {
                    column_id: col_id,
                    device_id: device.id,
                });
                total_cost += cost;
                break;
            }
        }
    }

    placements
}

pub fn place_data(strategy: PlacementStrategy, metadata: &Metadata, budget: Option<f32>) -> Vec<Placement> {
    match strategy {
        PlacementStrategy::HOT_TABLE => hot_table_strategy(metadata),
        PlacementStrategy::HOT_COLUMN => hot_column_strategy(metadata),
        PlacementStrategy::LOPT => lopt_strategy(metadata, budget),
    }
}

pub fn predict_scan_time(placements: &[Placement], table_scan: &TraceEntry, metadata: &Metadata) -> f32 {
    let mut device_times: HashMap<u32, f32> = HashMap::new();

    for &col_id in &table_scan.column_ids {
        if let Some(placement) = placements.iter().find(|p| p.column_id == col_id) {
            if let (Some(device), Some(device_model), Some(column)) = (
                metadata.get_device(placement.device_id),
                metadata.get_device_model(placement.device_id),
                metadata.get_column(col_id),
            ) {
                let compression_throughput = match device.compression {
                    crate::metadata::CompressionAlgorithm::None => device_model.none_throughput,
                    crate::metadata::CompressionAlgorithm::LZ4 => device_model.lz4_throughput,
                    crate::metadata::CompressionAlgorithm::ZSTD => device_model.zstd_throughput,
                };

                let column_time = device_model.seek_time + (column.size as f32 / compression_throughput);
                *device_times.entry(placement.device_id).or_insert(0.0) += column_time;
            }
        }
    }

    *device_times.values().max_by(|a, b| a.partial_cmp(b).unwrap()).unwrap_or(&0.0)
}

pub fn predict_total_time(placements: &[Placement], metadata: &Metadata) -> f32 {
    metadata.trace.iter().map(|trace| predict_scan_time(placements, trace, metadata)).sum()
}

pub fn calculate_total_cost(placements: &[Placement], metadata: &Metadata) -> f32 {
    placements.iter().map(|p| {
        if let (Some(device), Some(column)) = (metadata.get_device(p.device_id), metadata.get_column(p.column_id)) {
            device.cost_per_gb * (column.size as f32 / (1024.0 * 1024.0 * 1024.0))
        } else {
            0.0
        }
    }).sum()
}

pub fn optimize_placement(budget: f32, metadata: &Metadata) -> Vec<Placement> {
    let placements = place_data(PlacementStrategy::LOPT, metadata, Some(budget));
    let total_time = predict_total_time(&placements, metadata);
    let total_cost = calculate_total_cost(&placements, metadata);
    println!("Predicted total time: {:.2} s", total_time);
    println!("Total cost: {:.2}", total_cost);
    placements
}

pub fn apply_placements(placements: &[Placement], metadata: &Metadata, kv_store: &mut KVStore, data_retriever: &mut DataRetriever) {
    let mut chunks_to_move = Vec::new();

    // Collect all chunks that need to be moved
    for placement in placements {
        if let (Some(column), Some(target_device)) = (metadata.get_column(placement.column_id), metadata.get_device(placement.device_id)) {
            if let Some(chunks) = data_retriever.column_chunks.get(&placement.column_id) {
                for chunk in chunks {
                    if chunk.device_id != placement.device_id {
                        chunks_to_move.push((chunk.clone(), placement.device_id, column.name.clone(), target_device.name.clone()));
                    }
                }
            }
        }
    }

    // Move the collected chunks
    for (chunk, target_device_id, column_name, device_name) in chunks_to_move {
        data_retriever.move_column_chunk(&chunk, target_device_id, metadata, kv_store);
        println!("Moved column {} to device {}", column_name, device_name);
    }
}